compiler/rustc_typeck/src/astconv/generics.rs

```diff
@@ -43,23 +44,59 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             }
         }
 
+        let add_braces_suggestion = |arg: &GenericArg<'_>, err: &mut DiagnosticBuilder<'_>| {
+            let suggestions = vec![
+                (arg.span().shrink_to_lo(), String::from("{ ")),
+                (arg.span().shrink_to_hi(), String::from(" }")),
+            ];
+            err.multipart_suggestion(
+                "if this generic argument was intended as a const parameter, \
+                   surround it with braces:",
+                suggestions,
+                Applicability::MaybeIncorrect,
+            );
+        };
+
         // Specific suggestion set for diagnostics
         match (arg, &param.kind) {
             (
-                GenericArg::Type(hir::Ty { kind: hir::TyKind::Path { .. }, .. }),
-                GenericParamDefKind::Const { .. },
-            ) => {
-                let suggestions = vec![
-                    (arg.span().shrink_to_lo(), String::from("{ ")),
-                    (arg.span().shrink_to_hi(), String::from(" }")),
-                ];
-                err.multipart_suggestion(
-                    "if this generic argument was intended as a const parameter, \
-                try surrounding it with braces:",
-                    suggestions,
-                    Applicability::MaybeIncorrect,
-                );
-            }
+                GenericArg::Type(hir::Ty {
+                    kind: hir::TyKind::Path(rustc_hir::QPath::Resolved(_, path)),
+                    ..
+                }),
+                GenericParamDefKind::Const,
+            ) => match path.res {
+                Res::Err => {
+                    add_braces_suggestion(arg, &mut err);
+                    err.set_primary_message(
+                        "unresolved item provided when a constant was expected",
+                    );
+                }
+                Res::Def(DefKind::TyParam, src_def_id) => (|| {
+                    let param_hir_id = match param.def_id.as_local() {
+                        Some(x) => tcx.hir().local_def_id_to_hir_id(x),
+                        None => return,
+                    };
+                    let param_name = tcx.hir().ty_param_name(param_hir_id);
+                    let param_type = tcx.type_of(param.def_id);
+                    let (appl, sugg_type) = if param_type.is_suggestable() {
+                        (Applicability::MaybeIncorrect, format!("{}", param_type))
+                    } else {
+                        (Applicability::HasPlaceholders, String::from("T"))
```

Comment:
We probably want to remove the whole "else" branch because at that point we don't know if we're suggesting the right thing. (We don't know what we have.)